File: backend/core/dlq.py
# ----- dead letter queue @ backend/core/dlq.py -----
import json
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

import redis

from backend.utils.config import config
from backend.utils.logger import logger

DLQ_KEY = "dlq:ingest"


def _get_redis() -> redis.Redis:
    redis_url = config.redis_url
    kwargs = {}
    if redis_url.startswith("rediss://"):
        redis_url += "?ssl_cert_reqs=CERT_NONE"
    return redis.Redis.from_url(redis_url)


def push_to_dlq(
    job_id: str, task_name: str, args: Dict[str, Any], error: str, retry_count: int
) -> None:
    """Push a permanently failed job to the dead letter queue."""
    entry = {
        "job_id": job_id,
        "task": task_name,
        "args": args,
        "error": error,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "retry_count": retry_count,
    }

    rds = _get_redis()
    rds.rpush(DLQ_KEY, json.dumps(entry))
    logger.warning("Pushed job %s to DLQ (retries: %d)", job_id, retry_count)


def pop_dlq_entries(count: Optional[int] = None) -> List[Dict[str, Any]]:
    """Pop and return entries from the DLQ. If count is None, pop all."""
    rds = _get_redis()
    entries = []

    if count is None:
        count = rds.llen(DLQ_KEY)

    for _ in range(count):
        raw = rds.lpop(DLQ_KEY)
        if raw is None:
            break
        entries.append(json.loads(raw))

    return entries


def get_dlq_count() -> int:
    """Return the number of entries in the DLQ."""
    rds = _get_redis()
    return rds.llen(DLQ_KEY)

File: backend/internal/jobs/dlq.go
// ----- dead letter queue (Redis list) @ backend/internal/jobs/dlq.go -----
package jobs

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/anomalyco/autolinks/internal/logger"
)

const dlqKey = "dlq:ingest"

// DLQEntry represents a failed job entry in the dead letter queue.
type DLQEntry struct {
	JobID      string                 `json:"job_id"`
	Task       string                 `json:"task"`
	Args       map[string]interface{} `json:"args"`
	Error      string                 `json:"error"`
	Timestamp  string                 `json:"timestamp"`
	RetryCount int                    `json:"retry_count"`
}

// PushToDLQ pushes a permanently failed job to the dead letter queue.
func PushToDLQ(jobID, taskName string, args map[string]interface{}, errMsg string, retryCount int) {
	rds := getRedis()
	if rds == nil {
		logger.Error("Cannot push to DLQ: redis not configured")
		return
	}

	entry := DLQEntry{
		JobID:      jobID,
		Task:       taskName,
		Args:       args,
		Error:      errMsg,
		Timestamp:  time.Now().UTC().Format(time.RFC3339),
		RetryCount: retryCount,
	}

	data, err := json.Marshal(entry)
	if err != nil {
		logger.Error("Failed to marshal DLQ entry: %s", err)
		return
	}

	ctx := context.Background()
	if err := rds.RPush(ctx, dlqKey, data).Err(); err != nil {
		logger.Error("Failed to push to DLQ: %s", err)
		return
	}

	logger.Warning("Pushed job %s to DLQ (retries: %d)", jobID, retryCount)
}

// PopDLQEntries pops and returns entries from the DLQ. If count is 0, pops all.
func PopDLQEntries(count int) []DLQEntry {
	rds := getRedis()
	if rds == nil {
		logger.Error("Cannot pop DLQ: redis not configured")
		return nil
	}

	ctx := context.Background()

	if count <= 0 {
		l, err := rds.LLen(ctx, dlqKey).Result()
		if err != nil {
			logger.Error("Failed to get DLQ length: %s", err)
			return nil
		}
		count = int(l)
	}

	var entries []DLQEntry
	for i := 0; i < count; i++ {
		raw, err := rds.LPop(ctx, dlqKey).Result()
		if err != nil {
			break
		}

		var entry DLQEntry
		if err := json.Unmarshal([]byte(raw), &entry); err != nil {
			logger.Error("Failed to unmarshal DLQ entry: %s", err)
			continue
		}
		entries = append(entries, entry)
	}

	return entries
}

// GetDLQCount returns the number of entries in the DLQ.
func GetDLQCount() (int64, error) {
	rds := getRedis()
	if rds == nil {
		return 0, fmt.Errorf("redis not configured")
	}

	ctx := context.Background()
	return rds.LLen(ctx, dlqKey).Result()
}
